### PR TITLE
Add `report_dashboard.last_viewed_at`

### LINKED ISSUE #44251: Add a `last_viewed_at` timestamp to dashboards

### PR DESCRIPTION
**Pull request by johnswanson**
Defaults to `NOW()`. Also, reset all values of `report_card.last_used_at` to `NOW()` and make it non-nullable with a default of `NOW()`. See comments in the migrations file for details.

Fixes #44251

### CODE REVIEW

File: resources/migrations/001_update_migrations.yaml
Comment (noahmoss) [inline resources/migrations/001_update_migrations.yaml:8256]: These (at least the non-sql ones) should probably have `preConditions` or else they'll fail the new lint rule when merging. Would recommend making them idempotent if possible

File: resources/migrations/001_update_migrations.yaml
Comment (noahmoss) [inline resources/migrations/001_update_migrations.yaml:8274]: Is this index just to make the subsequent migration more efficient or is there another reason for it? Just for my own understanding.

**Review by noahmoss — APPROVED** [review]
great migration tests 👏

**Review by calherries — DISMISSED** [review]
Be careful with how you're covering all the ways a dashboard card can be queried. There are a lot of endpoints across normal, public, and embedded dashboards. I can see three endpoints missing from this PR that should trigger the `:event/dashboard-queried`:
```
POST /api/dashboard/pivot/:id/dashcard/:id/card/:id/query
GET /api/public/dashboard/:uuid/dashcard/:dashcard-id/card/:card-id
GET /api/public/pivot/dashboard/:uuid/dashcard/:dashcard-id/card/:card-id
```

File: src/metabase/api/embed.clj
Comment (calherries) [inline src/metabase/api/embed.clj:127]: It's a bit surprising that `:event/dashboard-queried` is published in the `process-query-for-dashcard-with-signed-token` function while `:event/card-read` is published in each endpoint handler separately. There's pros and cons either way and I don't have a strong opinion on it, but I think it should be consistent.

File: src/metabase/events/view_log.clj
Comment (calherries) [inline src/metabase/events/view_log.clj:67]: The dashboard model has an `updated_at` column that is automatically updated by toucan with this hook. 
https://github.com/metabase/metabase/blob/2d7891b8ed9f4dc29d12afe37c666d625bf686fa/src/metabase/models/interface.clj#L440-L443

I'm not sure what this column is used for, but its meaning will change substantially if we update it every time the dashboard is viewed. I've avoided this issue previously by skipping model hooks with this trick:
https://github.com/metabase/metabase/blob/2d7891b8ed9f4dc29d12afe37c666d625bf686fa/src/metabase/events/view_log.clj#L21-L24

Maybe `updated_at` is not actually used for anything, so this is okay. But I think we should avoid using toucan hooks here for performance reasons anyway. It will trigger extra queries we can easily avoid here.

**Review by calherries — COMMENTED** [review]
How should this column be handled during serialization import / export?

Card's `:last_used_at` is excluded from export. I'm not sure that's correct but worth checking.
https://github.com/metabase/metabase/blob/2d7891b8ed9f4dc29d12afe37c666d625bf686fa/src/metabase/models/card.clj#L896

File: resources/migrations/001_update_migrations.yaml
Comment (qnkhuat) [reply resources/migrations/001_update_migrations.yaml:8274]: yea I'd like to understand more about the motivation here. 

IIRC view_log is unbounded, so this migration will take a lot of time, plus it'll create overhead for inserting, which we do quite frequently in a synchronous fashion.

File: resources/migrations/001_update_migrations.yaml
Comment (qnkhuat) [inline resources/migrations/001_update_migrations.yaml:8292]: maybe you already know this but just want to share an interesting finding I found recently https://metaboat.slack.com/archives/C0641E4PB9B/p1720015474515889

File: src/metabase/events/view_log.clj
Comment (qnkhuat) [inline src/metabase/events/view_log.clj:60]: not part of this PR but we're working on adding batch processing for `report_card.last_viewed_at, view_count` and alike in https://github.com/metabase/metabase/pull/45007

do you think we can batch processing this as well?

File: resources/migrations/001_update_migrations.yaml
Comment (qnkhuat) [inline resources/migrations/001_update_migrations.yaml:8397]: Are we sure we want to make this assumption? what if someone creates a dashboard pragmatically? how about serialization? how about dashboard duplication?

File: resources/migrations/001_update_migrations.yaml
Comment (johnswanson) [reply resources/migrations/001_update_migrations.yaml:8274]: So yeah, it is to make the migration more efficient. Testing on data from stats (2.8M records in the view_log, 1650 dashboards), I found that without the index, the subsequent migration took ~30s. Creating the index took 10s and sped up the subsequent migration to ~200ms. I informally tested a few inserts into the view log before/after and the additional overhead wasn't detectable beneath the random variation in timing. Of course, if we do want to avoid that overhead I'm happy to delete the index in a subsequent migration.

**Comment by johnswanson** [comment]
> Be careful with how you're covering all the ways a dashboard card can be queried. There are a lot of endpoints across normal, public, and embedded dashboards. I can see three endpoints missing from this PR that should trigger the `:event/dashboard-queried`:
> 
> ```
> POST /api/dashboard/pivot/:id/dashcard/:id/card/:id/query
> GET /api/public/dashboard/:uuid/dashcard/:dashcard-id/card/:card-id
> GET /api/public/pivot/dashboard/:uuid/dashcard/:dashcard-id/card/:card-id
> ```

Yikes - thank you!

File: resources/migrations/001_update_migrations.yaml
Comment (johnswanson) [reply resources/migrations/001_update_migrations.yaml:8396]: To be clear, this isn't intended as an audit event or anything like that. It's just a marker intended as a way to find old, stale dashboards. I think defaulting to `now()` makes sense given that - though you do make a good point regarding serialization, I think in an ideal world we would export the `last_viewed_at` of dashboards and the `last_used_at` of cards, so that you could still identify stale items after an import. 🤔

**Comment by calherries** [comment]
@johnswanson 

> We update the last_used_at for cards in the query processor

We do that in the query processor because we need to use it to get nested card usages. I'm not sure we need to update `report_dashboard.last_viewed_at` in the QP. The QP is already complicated enough and using middleware to do this is overkill. I do think it is more future-proof to remove publishing the dashboard-queried event from the top-level endpoint handlers and instead move it to `metabase.query-processor.dashboard/process-query-for-dashcard` or `metabase.query-processor.card/process-query-for-card`, which are used across all the endpoints and also for dashboard subscriptions.

File: resources/migrations/001_update_migrations.yaml
Comment (calherries) [reply resources/migrations/001_update_migrations.yaml:8396]: > so that you could still identify stale items after an import. 

The need to identify stale items after an import could be a genuine use case. Maybe during serialization import we should compare `last_viewed_at` from both the imported entity and the existing entity, and keep the maximum.

File: resources/migrations/001_update_migrations.yaml
Comment (johnswanson) [reply resources/migrations/001_update_migrations.yaml:8292]: Lol, yes - I also discovered this incredibly fun fact. I love writing migrations like this for 4 separate database types 😅

File: resources/migrations/001_update_migrations.yaml
Comment (calherries) [reply resources/migrations/001_update_migrations.yaml:8396]: Checked with arakaki and confirmed `now` makes sense [here](https://metaboat.slack.com/archives/C0641E4PB9B/p1720549646851889). Would be great to leave a comment explaining why

File: resources/migrations/001_update_migrations.yaml
Comment (calherries) [inline resources/migrations/001_update_migrations.yaml:8338]: This would be awesome if it worked but unfortunately taking `greatest(max(report_card.last_used_at), max(view_log.timestamp))` can result in some `last_viewed_at` values that are further back in the past than the dashboard was actually viewed.

One problem is `view_log` used to be populated in OSS instances, but we removed it from OSS in v50. And `last_used_at` was [pulled out from some versions of v50](https://github.com/metabase/metabase/pull/44502). That means this scenario is possible:
1. A instance is created on version 49, using OSS
2. Dashboard A was viewed on date X, creating a `view_log` entry. (There's no `last_used_at` because `report_card.last_used_at` was introduced in 50.)
3. The instance upgrades to version 50.7 where `last_used_at` is not populated.
4. Dashboard A is viewed again on date Y. No `view_log` entry is created or `last_used_at` and no `last_used_at` value is set.
5. The instance upgrades to version 51.0. The `report_dashboard.last_viewed_at` value for dashboard A is X, but the logically correct value is Y.

There's more scenarios too, which might be more likely:
1. An EE instance is on version 50.7 or any other version in 50 where `last_used_at` is not present but `view_log` is being added to.
2. Dashboard A was viewed on date X, creating a `view_log` entry.
3. The instance changes from EE to OSS.
4.  Dashboard A is viewed again on date Y. No `view_log` entry is created because it's OSS or `last_used_at` and no `last_used_at` value is set.
5. The instance upgrades to version 51.0. The `report_dashboard.last_viewed_at` value for dashboard A is X, but the logically correct value is Y.

There's also another scenario where an OSS instance upgrades from 50.6 to 50.7, and no `last_used_at`. I could go on.

I can't see a fool-proof workaround to these problems :( And because this will be used for auto-clean-up I think we'd prefer to be conservative. To avoid setting `last_used_at` further in the past than reality in all scenarios, I think it has to default to `now`.

File: resources/migrations/001_update_migrations.yaml
Comment (qnkhuat) [reply resources/migrations/001_update_migrations.yaml:8274]: 👍 to skip the index, I don't think it's worth the trouble here given how little difference it makes

File: resources/migrations/001_update_migrations.yaml
Comment (johnswanson) [reply resources/migrations/001_update_migrations.yaml:8338]: This would certainly make the migrations a lot easier and lower risk 😅

Comment (johnswanson) [reply resources/migrations/001_update_migrations.yaml:8338]: @calherries I realized that there's a more general issue here.

1. Say someone is currently on v50.6 (populating `report_card.last_used_at`)
2. They upgrade tomorrow to v50.7 (we stop populating the column)
3. In 7 months, they upgrade to v51 and we release the "Show me my stale items" feature

We'll happily tell them "all of your cards are stale!" because they have non-null `last_used_at` values from >6 months ago. I'm thinking about what our options are here.

File: src/metabase/events/view_log.clj
Comment (johnswanson) [reply src/metabase/events/view_log.clj:60]: Makes sense to me, thanks for implementing this. Would you want me to do that in this PR (waiting until after the other one is merged), or in a followup PR, or would you take that?

One slight concern is that if we stopped writing `last_used_at` or `last_viewed_at` entirely for a long period of time, we might recommend that people delete things they shouldn't. I know grouper is "best effort" without any guarantees. But I think it's probably still an acceptable tradeoff if the odds of dropping any particular job are low.

**Comment by johnswanson** [comment]
@calherries / @qnkhuat / @noahmoss note that this has changed a lot since you last saw it and gotten a lot simpler - though possibly I should just close this PR and create a new one given that it's changed so much.

Given what we realized [here](https://github.com/metabase/metabase/pull/44240#discussion_r1672777008), I decided to just:
- wipe `report_card.last_used_at` and reset it to `NOW()` across the board, making it non-nullable
- create `report_dashboard.last_viewed_at` with a default of `NOW()`, no fancy stuff with `view_log` or `last_used_at`

Then I moved the event publish to `metabase.query-processor.dashboard/process-query-for-dashcard`.

Hopefully that seems reasonable - let me know if you have any questions.

File: resources/migrations/001_update_migrations.yaml
Comment (calherries) [reply resources/migrations/001_update_migrations.yaml]: Yes exactly, that's the scenario I (loosely) described here:

> There's also another scenario where an OSS instance upgrades from 50.6 to 50.7, and last_used_at goes from being updated to not.

**Comment by calherries** [comment]
Sounds good. You don't have to do the first bullet - I already went ahead an fixed `report_card.last_used_at` yesterday. I'm backporting it to 50 too.

https://github.com/metabase/metabase/pull/45323

**Comment by calherries** [comment]
@johnswanson your recent changes solves [this](https://github.com/metabase/metabase/pull/44240#discussion_r1671364760) issue but we still need to exclude it from serialization, and I think you should wait for Ngoc's suggested mechanism for updating `last_viewed_at` in batches, i.e. https://github.com/metabase/metabase/pull/45007

**Comment by johnswanson** [comment]
@qnkhuat or @noahmoss would you have a chance to take another look at this now that the concerns mentioned above have been addressed?

- we're excluding `last_viewed_at` from serialization
- we're batching the updates with grouper
- we're setting `last_viewed_at` to `NOW()` without any complex migration
- we have a single entrypoint where we emit the event that writes the `last_viewed_at` whenever a dashboard is viewed

Let me know if you see any other issues!

File: test/metabase/api/dashboard_test.clj
Comment (camsaul) [inline test/metabase/api/dashboard_test.clj:110]: This is an old test pattern, the better way to do this sort of stuff now is to use `=?`

File: resources/migrations/001_update_migrations.yaml
Comment (camsaul) [inline resources/migrations/001_update_migrations.yaml:8376]: Pretty sure the rollback should get autogenerated for `addColumn` migrations?

File: src/metabase/events/view_log.clj
Comment (qnkhuat) [inline src/metabase/events/view_log.clj:99]: Setting all these dashboards to have `last_used_at = now` is inaccurate because `now` is the batch processing time.

We can be more accurate here if you record the `submitted` time of a dashboard and use a case statement to update it.

I did a similar thing in https://github.com/metabase/metabase/blob/1a07aa45ed7f1e9f1b7b137672855b8d578b73a1/src/metabase/query_processor/middleware/update_used_cards.clj#L23

**Review by qnkhuat — CHANGES_REQUESTED** [review]
I think this is heading to the right direction but:
- Currently `last_viewed_at` is set to the gouper batch processing time, we could be more accurate and set it to submitted time using a case clause.
- we need some tests for this. I think we could have 2 layers of test here:
  - QP tests for `process-query-for-dashcard` to update `last_viewed_at`
  - API tests: private, public, embed apis to get dashcard result should update last_viewed_at
  
  
Tip: you can use `(mt/with-temporary-setting-values [synchronous-batch-updates true])` to for grouper submit to process synchronously during test.

File: src/metabase/events/view_log.clj
Comment (johnswanson) [reply src/metabase/events/view_log.clj:99]: I thought about this - are you sure it's worth the additional complexity to be precise here?

Maybe we should change the name of the field to `:not_used_after`/`:not_viewed_after`, but I think that we can not worry too much about being ~20s later than reality. The times are best effort anyway - ideally we'll err on the later side, but even that isn't always possible:
- the current `:last_viewed_at` may be whatever time the migration was run to create this column
- it may be one of the times the card/dashboard was used, but some subsequent usages were lost

It seems to me that the times here are approximate and that's ok - as far as I know the only planned use of this column is the "cleanup stale items" project, and if we don't cleanup something that was used 6 months minus 20 seconds ago when we should have, I don't think anyone will notice or complain 😅

Comment (qnkhuat) [reply src/metabase/events/view_log.clj]: hmm, sorry but I don't see the downside of doing that, it's quite simple IMO.

Also we currently set the batch process time to 20s but there is nothing stopping us from increasing that in the future.

Comment (johnswanson) [reply src/metabase/events/view_log.clj]: Ok, that makes sense - done

File: src/metabase/events/view_log.clj
Comment (qnkhuat) [inline src/metabase/events/view_log.clj:109]: > Failed to update dashboard last_used_at

File: src/metabase/events/view_log.clj
Comment (qnkhuat) [inline src/metabase/events/view_log.clj:99]: do we have tests for this?

I have written one for update last_used_at here https://github.com/metabase/metabase/blob/24d47437cacec9dece4cd50921d29c7a26183c4f/test/metabase/query_processor/middleware/update_used_cards_test.clj#L73

File: test/metabase/api/embed_test.clj
Comment (qnkhuat) [inline test/metabase/api/embed_test.clj:1672]: the indent is a bit off here

File: src/metabase/events/view_log.clj
Comment (qnkhuat) [inline src/metabase/events/view_log.clj:109]:
```suggestion
        (log/error e "Failed to update dashboard last_viewed_at")))))
```

File: test/metabase/query_processor/dashboard_test.clj
Comment (qnkhuat) [inline test/metabase/query_processor/dashboard_test.clj:410]:
```suggestion
  (mt/test-helpers-set-global-values!
```

I don't think that dynamic var is meant to be bound manually, we should use the macro instead

File: src/metabase/events/view_log.clj
Comment (johnswanson) [reply src/metabase/events/view_log.clj:99]: Good callout, just added a test for this.

File: test/metabase/query_processor/dashboard_test.clj
Comment (johnswanson) [reply test/metabase/query_processor/dashboard_test.clj]: Oops, you're right - thanks!

**Review by qnkhuat — APPROVED** [review]
It’s LGTM now, thanks for addressing my comments !

**Pull request merged**
